File: nodejs/muse/util/Sequences.js
"use strict"

const QuickSort     = require("muse/algorithms/sorting/QuickSort")
const RandomFactory = require("muse/util/RandomFactory")

exports.inspect = (arr) => {
    console.log("[")
    for (let i = 0, length = arr.length; i < length; i++) {
        process.stdout.write(`\t#${i.toString(16).toUpperCase().padStart(4, "0")}  -->  ${arr[i]}\n`)
    }
    console.log("]")
}

exports.isSorted = (arr) => {
    for (let i = 1, length = arr.length; i < length; i++) {
        if (arr[i] < arr[i - 1]) {
            return false
        }
    }

    return true
}

exports.parityChecksum = (arr) => {
    let checksum = 0

    for (let v of arr) {
        checksum ^= v
    }

    return checksum
}

exports.reverse = (arr) => {
    let k
    let temp

    for (let i = 0, bound = Math.floor(arr.length >>> 1), length = arr.length; i < bound; i++) {
        k = length - i - 1
        temp = arr[i]
        arr[i] = arr[k]
        arr[k] = temp
    }
}

exports.shuffle = (arr) => {
    let k
    let temp

    RandomFactory.seed()
    for (let i = 0, length = arr.length; i < length; i++) {
        k = RandomFactory.genIntN(i, length)
        temp = arr[i]
        arr[i] = arr[k]
        arr[k] = temp
    }
}

exports.sort = (arr) => {
    QuickSort.sort(arr)
}

File: python/muse/util/sequences.py
from muse.algorithms.sorting import quick_sort
from muse.util import random_factory


def inspect(arr: list[int]) -> None:
    print("[")
    for i, v in enumerate(arr):
        print("\t#%04X  -->  %d" % (i, v))
    print("]")


def is_sorted(arr: list[int]) -> bool:
    for i in range(1, len(arr)):
        if arr[i] < arr[i - 1]:
            return False

    return True


def parity_checksum(arr: list[int]) -> int:
    checksum = 0

    for v in arr:
        checksum ^= v

    return checksum


def reverse(arr: list[int]) -> None:
    for i in range(len(arr) >> 1):
        k = len(arr) - i - 1
        arr[i], arr[k] = arr[k], arr[i]


def shuffle(arr: list[int]) -> None:
    for i in range(len(arr)):
        k = random_factory.gen_int_n(i, len(arr) - 1)
        arr[i], arr[k] = arr[k], arr[i]


def sort(arr: list[int]) -> None:
    quick_sort.sort(arr)
